### PR TITLE
[NPU] Add sync_batch_norm and sync_batch_norm_grad NPU Kernel

### PR DESCRIPTION
**Pull request by windstamp**
### PR types

New features
### PR changes

OPs
### Describe

[NPU] Add sync_batch_norm and sync_batch_norm_grad NPU Kernel

50d5f5970df1a8e8813d39943c6d534f

1c2a412ce222cd58927c6c4a6122631c

dff24f0ca5e66900b9743b81a7e00ea3

### CODE REVIEW

File: paddle/fluid/operators/batch_norm_op_npu.cc [hunk 1]
```diff
@@ -76,10 +76,10 @@ class NPUBatchNormOpKernel : public framework::OpKernel<T> {
...
       auto *saved_mean = ctx.Output<Tensor>("SavedMean");
       auto *saved_variance = ctx.Output<Tensor>("SavedVariance");
-      mean_out->mutable_data<T>(ctx.GetPlace());
-      variance_out->mutable_data<T>(ctx.GetPlace());
-      saved_mean->mutable_data<T>(ctx.GetPlace());
-      saved_variance->mutable_data<T>(ctx.GetPlace());
+      mean_out->mutable_data<float>(ctx.GetPlace());
+      variance_out->mutable_data<float>(ctx.GetPlace());
+      saved_mean->mutable_data<float>(ctx.GetPlace());
+      saved_variance->mutable_data<float>(ctx.GetPlace());
```
Comment (qili93) [inline paddle/fluid/operators/batch_norm_op_npu.cc]: why change the datatype to float, the template datatype T should also equal to float16?

File: paddle/fluid/operators/batch_norm_op_npu.cc [hunk 2]
```diff
@@ -169,8 +169,8 @@ class NPUBatchNormGradOpKernel : public framework::OpKernel<T> {
 
     auto stream = ctx.template device_context<NPUDeviceContext>().stream();
     if (d_scale && d_bias) {
-      d_scale->mutable_data<T>(ctx.GetPlace());
-      d_bias->mutable_data<T>(ctx.GetPlace());
+      d_scale->mutable_data<float>(ctx.GetPlace());
+      d_bias->mutable_data<float>(ctx.GetPlace());
```
Comment (qili93) [inline paddle/fluid/operators/batch_norm_op_npu.cc]: same question as above.

File: paddle/fluid/operators/conv_op_npu.cc [hunk 3]
```diff
@@ -186,6 +186,11 @@ class DepthwiseConvGradNPUKernel : public framework::OpKernel<T> {
       dilations[3] = dilation[1];
     }
 
+    // LOG(INFO) << "strides = " << framework::make_ddim(strides).to_str();
+    // LOG(INFO) << "dilations = " << framework::make_ddim(dilations).to_str();
+    // LOG(INFO) << "padding = " << framework::make_ddim(padding).to_str();
+    // LOG(INFO) << "data_format = " << data_format;
+
```
Comment (qili93) [inline paddle/fluid/operators/conv_op_npu.cc]: remove the above lines.

File: paddle/fluid/operators/conv_op_npu.cc [hunk 4]
```diff
@@ -302,12 +307,19 @@ template <typename T>
...
     auto input = ctx.Input<Tensor>("Input");
     auto filter = ctx.Input<Tensor>("Filter");
     auto output_grad = ctx.Input<Tensor>(framework::GradVarName("Output"));
     auto input_grad = ctx.Output<Tensor>(framework::GradVarName("Input"));
     auto filter_grad = ctx.Output<Tensor>(framework::GradVarName("Filter"));
 
+    LOG(WARNING) << "input->type(): " << input->type();
+    LOG(WARNING) << "filter->type(): " << filter->type();
+    LOG(WARNING) << "output_grad->type(): " << output_grad->type();
+
```
Comment (qili93) [inline paddle/fluid/operators/conv_op_npu.cc]: remove these LOG(WARNING) lines.

File: paddle/fluid/operators/sync_batch_norm_op_npu.cc [hunk 5]
```diff
@@ -0,0 +1,1787 @@
...
+See the License for the specific language governing permissions and
+limitations under the Licnse. */
+
+#include "paddle/fluid/operators/batch_norm_op.h"
+#include "paddle/fluid/operators/npu_op_runner.h"
+
+#if defined(PADDLE_WITH_ASCEND_CL)
+#include "paddle/fluid/platform/collective_helper.h"
+#include "paddle/fluid/platform/hccl_helper.h"
+#endif
```
Comment (qili93) [inline paddle/fluid/operators/sync_batch_norm_op_npu.cc]: why #if defined(PADDLE_WITH_ASCEND_CL) here? this npu.cc file will not be compiled if WITH_ASCEND OFF.

File: paddle/fluid/operators/sync_batch_norm_op_npu.cc [hunk 6]
```diff
@@ -0,0 +1,1787 @@
...
+                 const framework::ExecutionContext &ctx) {
+  LOG(WARNING) << "shape: " << src.dims();
+  // return;
+
+  std::vector<T> vec(src.numel());
+  TensorToVector(src, ctx.device_context(), &vec);
+  ctx.template device_context<paddle::platform::NPUDeviceContext>().Wait();
+  LOG(WARNING) << "vec: " << outputVector<T>(vec);
+}
+
```
Comment (qili93) [inline paddle/fluid/operators/sync_batch_norm_op_npu.cc]: remove debug functions of outputVector and PrintTensor here

File: paddle/fluid/operators/sync_batch_norm_op_npu.cc [hunk 7]
```diff
@@ -0,0 +1,1787 @@
...
+  else if (layout == framework::DataLayout::kNHWC)
+    multiples = {N, H, W, 1};
+
+  Tensor common_mean_tile_1;
+  {
+    common_mean_tile_1.Resize({C});
+    common_mean_tile_1.mutable_data<float>(place);
+
+    TensorCopySync(*common_mean, place, &common_mean_tile_1);
+    LOG(WARNING) << "common_mean_tile_1: ";
```
Comment (qili93) [inline paddle/fluid/operators/sync_batch_norm_op_npu.cc]: remove all LOG(WARNING) messages.

**Review by qili93 — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/batch_norm_op_npu.cc (hunk 1 above)
Comment (windstamp) [reply paddle/fluid/operators/batch_norm_op_npu.cc]: thanks.

1. 参考 batch_norm_op.cu，参数 X 和 Y 为 float16 时，其它参数（mean, variance, scale, bias 等）仍然是 float32。
2. 背后的原因可能是，对于模型训练（比如 AMP - float16 训练），只是会将参数 X 和 Y 转换为 float16，其它的参数仍然是 float32，所以 batch_norm_op.cu 是目前的实现。
3. 类似于第 2 点，单元测试构建的模型，参数 mean, variance, scale, bias 等即使对于 float16 测试 case，仍然是 float32。
4. Ascend BNTrainingUpdate 的文档，对于上述参数类型的描述是类似的。

c0fdd206418021cab49c777799536a7d

File: paddle/fluid/operators/batch_norm_op_npu.cc (hunk 2 above)
Comment (windstamp) [reply paddle/fluid/operators/batch_norm_op_npu.cc]: thanks.

解释如上。

File: paddle/fluid/operators/conv_op_npu.cc (hunk 3 above)
Comment (windstamp) [reply paddle/fluid/operators/conv_op_npu.cc]: done, thanks.

File: paddle/fluid/operators/conv_op_npu.cc (hunk 4 above)
Comment (windstamp) [reply paddle/fluid/operators/conv_op_npu.cc]: done, thanks.

File: paddle/fluid/operators/sync_batch_norm_op_npu.cc (hunk 5 above)
Comment (windstamp) [reply paddle/fluid/operators/sync_batch_norm_op_npu.cc]: Deleted, thanks.

File: paddle/fluid/operators/sync_batch_norm_op_npu.cc (hunk 6 above)
Comment (windstamp) [reply paddle/fluid/operators/sync_batch_norm_op_npu.cc]: done, thanks.

File: paddle/fluid/operators/sync_batch_norm_op_npu.cc (hunk 7 above)
Comment (windstamp) [reply paddle/fluid/operators/sync_batch_norm_op_npu.cc]: done, thanks.

**Review by qili93 — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**